// virtual-list-down-up/dist/index.js
import './style.css';
import Le, { useState as ee, useRef as $, useEffect as Ze, useLayoutEffect as le, useImperativeHandle as Rr } from "react";
var Ae = { exports: {} }, ge = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Ke;
function Er() {
  if (Ke) return ge;
  Ke = 1;
  var u = Le, h = Symbol.for("react.element"), w = Symbol.for("react.fragment"), f = Object.prototype.hasOwnProperty, g = u.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, v = { key: !0, ref: !0, __self: !0, __source: !0 };
  function m(F, R, b) {
    var y, C = {}, P = null, O = null;
    b !== void 0 && (P = "" + b), R.key !== void 0 && (P = "" + R.key), R.ref !== void 0 && (O = R.ref);
    for (y in R) f.call(R, y) && !v.hasOwnProperty(y) && (C[y] = R[y]);
    if (F && F.defaultProps) for (y in R = F.defaultProps, R) C[y] === void 0 && (C[y] = R[y]);
    return { $$typeof: h, type: F, key: P, ref: O, props: C, _owner: g.current };
  }
  return ge.Fragment = w, ge.jsx = m, ge.jsxs = m, ge;
}
var me = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Ge;
function _r() {
  return Ge || (Ge = 1, process.env.NODE_ENV !== "production" && function() {
    var u = Le, h = Symbol.for("react.element"), w = Symbol.for("react.portal"), f = Symbol.for("react.fragment"), g = Symbol.for("react.strict_mode"), v = Symbol.for("react.profiler"), m = Symbol.for("react.provider"), F = Symbol.for("react.context"), R = Symbol.for("react.forward_ref"), b = Symbol.for("react.suspense"), y = Symbol.for("react.suspense_list"), C = Symbol.for("react.memo"), P = Symbol.for("react.lazy"), O = Symbol.for("react.offscreen"), k = Symbol.iterator, q = "@@iterator";
    function x(e) {
      if (e === null || typeof e != "object")
        return null;
      var r = k && e[k] || e[q];
      return typeof r == "function" ? r : null;
    }
    var L = u.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function p(e) {
      {
        for (var r = arguments.length, t = new Array(r > 1 ? r - 1 : 0), n = 1; n < r; n++)
          t[n - 1] = arguments[n];
        j("error", e, t);
      }
    }
    function j(e, r, t) {
      {
        var n = L.ReactDebugCurrentFrame, c = n.getStackAddendum();
        c !== "" && (r += "%s", t = t.concat([c]));
        var d = t.map(function(l) {
          return String(l);
        });
        d.unshift("Warning: " + r), Function.prototype.apply.call(console[e], console, d);
      }
    }
    var re = !1, K = !1, G = !1, I = !1, z = !1, be;
    be = Symbol.for("react.module.reference");
    function B(e) {
      return !!(typeof e == "string" || typeof e == "function" || e === f || e === v || z || e === g || e === b || e === y || I || e === O || re || K || G || typeof e == "object" && e !== null && (e.$$typeof === P || e.$$typeof === C || e.$$typeof === m || e.$$typeof === F || e.$$typeof === R || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      e.$$typeof === be || e.getModuleId !== void 0));
    }
    function ye(e, r, t) {
      var n = e.displayName;
      if (n)
        return n;
      var c = r.displayName || r.name || "";
      return c !== "" ? t + "(" + c + ")" : t;
    }
    function te(e) {
      return e.displayName || "Context";
    }
    function M(e) {
      if (e == null)
        return null;
      if (typeof e.tag == "number" && p("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof e == "function")
        return e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case f:
          return "Fragment";
        case w:
          return "Portal";
        case v:
          return "Profiler";
        case g:
          return "StrictMode";
        case b:
          return "Suspense";
        case y:
          return "SuspenseList";
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case F:
            var r = e;
            return te(r) + ".Consumer";
          case m:
            var t = e;
            return te(t._context) + ".Provider";
          case R:
            return ye(e, e.render, "ForwardRef");
          case C:
            var n = e.displayName || null;
            return n !== null ? n : M(e.type) || "Memo";
          case P: {
            var c = e, d = c._payload, l = c._init;
            try {
              return M(l(d));
            } catch {
              return null;
            }
          }
        }
      return null;
    }
    var W = Object.assign, J = 0, ce, ne, Re, Ee, _e, X, Se;
    function Te() {
    }
    Te.__reactDisabledLog = !0;
    function we() {
      {
        if (J === 0) {
          ce = console.log, ne = console.info, Re = console.warn, Ee = console.error, _e = console.group, X = console.groupCollapsed, Se = console.groupEnd;
          var e = {
            configurable: !0,
            enumerable: !0,
            value: Te,
            writable: !0
          };
          Object.defineProperties(console, {
            info: e,
            log: e,
            warn: e,
            error: e,
            group: e,
            groupCollapsed: e,
            groupEnd: e
          });
        }
        J++;
      }
    }
    function xe() {
      {
        if (J--, J === 0) {
          var e = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: W({}, e, {
              value: ce
            }),
            info: W({}, e, {
              value: ne
            }),
            warn: W({}, e, {
              value: Re
            }),
            error: W({}, e, {
              value: Ee
            }),
            group: W({}, e, {
              value: _e
            }),
            groupCollapsed: W({}, e, {
              value: X
            }),
            groupEnd: W({}, e, {
              value: Se
            })
          });
        }
        J < 0 && p("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var ue = L.ReactCurrentDispatcher, fe;
    function oe(e, r, t) {
      {
        if (fe === void 0)
          try {
            throw Error();
          } catch (c) {
            var n = c.stack.trim().match(/\n( *(at )?)/);
            fe = n && n[1] || "";
          }
        return `
` + fe + e;
      }
    }
    var de = !1, Z;
    {
      var je = typeof WeakMap == "function" ? WeakMap : Map;
      Z = new je();
    }
    function a(e, r) {
      if (!e || de)
        return "";
      {
        var t = Z.get(e);
        if (t !== void 0)
          return t;
      }
      var n;
      de = !0;
      var c = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var d;
      d = ue.current, ue.current = null, we();
      try {
        if (r) {
          var l = function() {
            throw Error();
          };
          if (Object.defineProperty(l.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(l, []);
            } catch (A) {
              n = A;
            }
            Reflect.construct(e, [], l);
          } else {
            try {
              l.call();
            } catch (A) {
              n = A;
            }
            e.call(l.prototype);
          }
        } else {
          try {
            throw Error();
          } catch (A) {
            n = A;
          }
          e();
        }
      } catch (A) {
        if (A && n && typeof A.stack == "string") {
          for (var i = A.stack.split(`
`), D = n.stack.split(`
`), E = i.length - 1, T = D.length - 1; E >= 1 && T >= 0 && i[E] !== D[T]; )
            T--;
          for (; E >= 1 && T >= 0; E--, T--)
            if (i[E] !== D[T]) {
              if (E !== 1 || T !== 1)
                do
                  if (E--, T--, T < 0 || i[E] !== D[T]) {
                    var N = `
` + i[E].replace(" at new ", " at ");
                    return e.displayName && N.includes("<anonymous>") && (N = N.replace("<anonymous>", e.displayName)), typeof e == "function" && Z.set(e, N), N;
                  }
                while (E >= 1 && T >= 0);
              break;
            }
        }
      } finally {
        de = !1, ue.current = d, xe(), Error.prepareStackTrace = c;
      }
      var ae = e ? e.displayName || e.name : "", Q = ae ? oe(ae) : "";
      return typeof e == "function" && Z.set(e, Q), Q;
    }
    function o(e, r, t) {
      return a(e, !1);
    }
    function S(e) {
      var r = e.prototype;
      return !!(r && r.isReactComponent);
    }
    function s(e, r, t) {
      if (e == null)
        return "";
      if (typeof e == "function")
        return a(e, S(e));
      if (typeof e == "string")
        return oe(e);
      switch (e) {
        case b:
          return oe("Suspense");
        case y:
          return oe("SuspenseList");
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case R:
            return o(e.render);
          case C:
            return s(e.type, r, t);
          case P: {
            var n = e, c = n._payload, d = n._init;
            try {
              return s(d(c), r, t);
            } catch {
            }
          }
        }
      return "";
    }
    var V = Object.prototype.hasOwnProperty, Y = {}, ve = L.ReactDebugCurrentFrame;
    function H(e) {
      if (e) {
        var r = e._owner, t = s(e.type, e._source, r ? r.type : null);
        ve.setExtraStackFrame(t);
      } else
        ve.setExtraStackFrame(null);
    }
    function Me(e, r, t, n, c) {
      {
        var d = Function.call.bind(V);
        for (var l in e)
          if (d(e, l)) {
            var i = void 0;
            try {
              if (typeof e[l] != "function") {
                var D = Error((n || "React class") + ": " + t + " type `" + l + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof e[l] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw D.name = "Invariant Violation", D;
              }
              i = e[l](r, l, n, t, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (E) {
              i = E;
            }
            i && !(i instanceof Error) && (H(c), p("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", n || "React class", t, l, typeof i), H(null)), i instanceof Error && !(i.message in Y) && (Y[i.message] = !0, H(c), p("Failed %s type: %s", t, i.message), H(null));
          }
      }
    }
    var pe = Array.isArray;
    function U(e) {
      return pe(e);
    }
    function er(e) {
      {
        var r = typeof Symbol == "function" && Symbol.toStringTag, t = r && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return t;
      }
    }
    function rr(e) {
      try {
        return Ye(e), !1;
      } catch {
        return !0;
      }
    }
    function Ye(e) {
      return "" + e;
    }
    function Ne(e) {
      if (rr(e))
        return p("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", er(e)), Ye(e);
    }
    var he = L.ReactCurrentOwner, tr = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, $e, We, Oe;
    Oe = {};
    function nr(e) {
      if (V.call(e, "ref")) {
        var r = Object.getOwnPropertyDescriptor(e, "ref").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.ref !== void 0;
    }
    function or(e) {
      if (V.call(e, "key")) {
        var r = Object.getOwnPropertyDescriptor(e, "key").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function ir(e, r) {
      if (typeof e.ref == "string" && he.current && r && he.current.stateNode !== r) {
        var t = M(he.current.type);
        Oe[t] || (p('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref', M(he.current.type), e.ref), Oe[t] = !0);
      }
    }
    function ar(e, r) {
      {
        var t = function() {
          $e || ($e = !0, p("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        t.isReactWarning = !0, Object.defineProperty(e, "key", {
          get: t,
          configurable: !0
        });
      }
    }
    function sr(e, r) {
      {
        var t = function() {
          We || (We = !0, p("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        t.isReactWarning = !0, Object.defineProperty(e, "ref", {
          get: t,
          configurable: !0
        });
      }
    }
    var lr = function(e, r, t, n, c, d, l) {
      var i = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: h,
        // Built-in properties that belong on the element
        type: e,
        key: r,
        ref: t,
        props: l,
        // Record the component responsible for creating this element.
        _owner: d
      };
      return i._store = {}, Object.defineProperty(i._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(i, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: n
      }), Object.defineProperty(i, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: c
      }), Object.freeze && (Object.freeze(i.props), Object.freeze(i)), i;
    };
    function cr(e, r, t, n, c) {
      {
        var d, l = {}, i = null, D = null;
        t !== void 0 && (Ne(t), i = "" + t), or(r) && (Ne(r.key), i = "" + r.key), nr(r) && (D = r.ref, ir(r, c));
        for (d in r)
          V.call(r, d) && !tr.hasOwnProperty(d) && (l[d] = r[d]);
        if (e && e.defaultProps) {
          var E = e.defaultProps;
          for (d in E)
            l[d] === void 0 && (l[d] = E[d]);
        }
        if (i || D) {
          var T = typeof e == "function" ? e.displayName || e.name || "Unknown" : e;
          i && ar(l, T), D && sr(l, T);
        }
        return lr(e, i, D, c, n, he.current, l);
      }
    }
    var Ce = L.ReactCurrentOwner, Ve = L.ReactDebugCurrentFrame;
    function ie(e) {
      if (e) {
        var r = e._owner, t = s(e.type, e._source, r ? r.type : null);
        Ve.setExtraStackFrame(t);
      } else
        Ve.setExtraStackFrame(null);
    }
    var Pe;
    Pe = !1;
    function ke(e) {
      return typeof e == "object" && e !== null && e.$$typeof === h;
    }
    function He() {
      {
        if (Ce.current) {
          var e = M(Ce.current.type);
          if (e)
            return `

Check the render method of \`` + e + "`.";
        }
        return "";
      }
    }
    function ur(e) {
      return "";
    }
    var Ue = {};
    function fr(e) {
      {
        var r = He();
        if (!r) {
          var t = typeof e == "string" ? e : e.displayName || e.name;
          t && (r = `

Check the top-level render call using <` + t + ">.");
        }
        return r;
      }
    }
    function ze(e, r) {
      {
        if (!e._store || e._store.validated || e.key != null)
          return;
        e._store.validated = !0;
        var t = fr(r);
        if (Ue[t])
          return;
        Ue[t] = !0;
        var n = "";
        e && e._owner && e._owner !== Ce.current && (n = " It was passed a child from " + M(e._owner.type) + "."), ie(e), p('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', t, n), ie(null);
      }
    }
    function Be(e, r) {
      {
        if (typeof e != "object")
          return;
        if (U(e))
          for (var t = 0; t < e.length; t++) {
            var n = e[t];
            ke(n) && ze(n, r);
          }
        else if (ke(e))
          e._store && (e._store.validated = !0);
        else if (e) {
          var c = x(e);
          if (typeof c == "function" && c !== e.entries)
            for (var d = c.call(e), l; !(l = d.next()).done; )
              ke(l.value) && ze(l.value, r);
        }
      }
    }
    function dr(e) {
      {
        var r = e.type;
        if (r == null || typeof r == "string")
          return;
        var t;
        if (typeof r == "function")
          t = r.propTypes;
        else if (typeof r == "object" && (r.$$typeof === R || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        r.$$typeof === C))
          t = r.propTypes;
        else
          return;
        if (t) {
          var n = M(r);
          Me(t, e.props, "prop", n, e);
        } else if (r.PropTypes !== void 0 && !Pe) {
          Pe = !0;
          var c = M(r);
          p("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", c || "Unknown");
        }
        typeof r.getDefaultProps == "function" && !r.getDefaultProps.isReactClassApproved && p("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function vr(e) {
      {
        for (var r = Object.keys(e.props), t = 0; t < r.length; t++) {
          var n = r[t];
          if (n !== "children" && n !== "key") {
            ie(e), p("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", n), ie(null);
            break;
          }
        }
        e.ref !== null && (ie(e), p("Invalid attribute `ref` supplied to `React.Fragment`."), ie(null));
      }
    }
    var Je = {};
    function qe(e, r, t, n, c, d) {
      {
        var l = B(e);
        if (!l) {
          var i = "";
          (e === void 0 || typeof e == "object" && e !== null && Object.keys(e).length === 0) && (i += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var D = ur();
          D ? i += D : i += He();
          var E;
          e === null ? E = "null" : U(e) ? E = "array" : e !== void 0 && e.$$typeof === h ? (E = "<" + (M(e.type) || "Unknown") + " />", i = " Did you accidentally export a JSX literal instead of a component?") : E = typeof e, p("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", E, i);
        }
        var T = cr(e, r, t, c, d);
        if (T == null)
          return T;
        if (l) {
          var N = r.children;
          if (N !== void 0)
            if (n)
              if (U(N)) {
                for (var ae = 0; ae < N.length; ae++)
                  Be(N[ae], e);
                Object.freeze && Object.freeze(N);
              } else
                p("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              Be(N, e);
        }
        if (V.call(r, "key")) {
          var Q = M(e), A = Object.keys(r).filter(function(yr) {
            return yr !== "key";
          }), Ie = A.length > 0 ? "{key: someKey, " + A.join(": ..., ") + ": ...}" : "{key: someKey}";
          if (!Je[Q + Ie]) {
            var br = A.length > 0 ? "{" + A.join(": ..., ") + ": ...}" : "{}";
            p(`A props object containing a "key" prop is being spread into JSX:
  let props = %s;
  <%s {...props} />
React keys must be passed directly to JSX without using spread:
  let props = %s;
  <%s key={someKey} {...props} />`, Ie, Q, br, Q), Je[Q + Ie] = !0;
          }
        }
        return e === f ? vr(T) : dr(T), T;
      }
    }
    function pr(e, r, t) {
      return qe(e, r, t, !0);
    }
    function hr(e, r, t) {
      return qe(e, r, t, !1);
    }
    var gr = hr, mr = pr;
    me.Fragment = f, me.jsx = gr, me.jsxs = mr;
  }()), me;
}
process.env.NODE_ENV === "production" ? Ae.exports = Er() : Ae.exports = _r();
var _ = Ae.exports;
const De = new Array(100).fill(!0).map((u, h) => ({
  id: h,
  value: h
})), se = 40, Cr = () => {
  const [u, h] = ee(0), [w, f] = ee(0), [g, v] = ee(0), m = $(null), F = De.length * se, R = Math.ceil(u / se), b = g + R, y = De.slice(g, Math.min(b, De.length)), C = `translate3d(0, ${w}px, 0)`;
  Ze(() => {
    h(m.current.clientHeight);
  }, []);
  const P = () => {
    const O = m.current.scrollTop;
    v(Math.floor(O / se)), f(O - O % se);
  };
  return /* @__PURE__ */ _.jsxs(
    "div",
    {
      ref: m,
      className: "infinite-fixed-list-container",
      onScroll: P,
      children: [
        /* @__PURE__ */ _.jsx(
          "div",
          {
            className: "infinite-fixed-list-phantom",
            style: { height: F + "px" }
          }
        ),
        /* @__PURE__ */ _.jsx("div", { className: "infinite-fixed-list", style: { transform: C }, children: y.map((O) => /* @__PURE__ */ _.jsx(
          "div",
          {
            className: "infinite-fixed-list-item",
            style: { height: se + "px", lineHeight: se + "px" },
            children: O.value
          },
          O.id
        )) })
      ]
    }
  );
}, Qe = ({ hasMore: u }) => u && /* @__PURE__ */ _.jsxs("div", { className: "infinite-loading", children: [
  /* @__PURE__ */ _.jsx("div", { className: "infinite-loading-dot" }),
  /* @__PURE__ */ _.jsx("div", { className: "infinite-loading-dot" }),
  /* @__PURE__ */ _.jsx("div", { className: "infinite-loading-dot" }),
  /* @__PURE__ */ _.jsx("div", { className: "infinite-loading-dot" }),
  /* @__PURE__ */ _.jsx("div", { className: "infinite-loading-dot" })
] }), Sr = ({
  refs: u,
  dataId: h,
  topLoadMoreCallback: w,
  hasMoreTopData: f,
  loaderAtTop: g
}) => (le(() => {
  if (!f) return;
  const v = new IntersectionObserver(
    (m) => {
      m[0].isIntersecting && w();
    },
    {
      root: null,
      rootMargin: "0px",
      threshold: 0
    }
  );
  return u.current && v.observe(u.current), () => {
    u.current && v.unobserve(u.current);
  };
}, [f]), /* @__PURE__ */ _.jsx(
  "div",
  {
    ref: u,
    className: "infinite-list-item infinite-list-pull-refesh",
    "data-id": h,
    children: g || /* @__PURE__ */ _.jsx(Qe, { hasMore: f })
  }
)), Tr = ({
  refs: u,
  dataId: h,
  bottomLoadMoreCallback: w,
  hasMoreBottomData: f,
  loaderAtBottom: g
}) => (le(() => {
  if (!f) return;
  const v = new IntersectionObserver(
    (m) => {
      m[0].isIntersecting && w();
    },
    {
      root: null,
      rootMargin: "0px",
      threshold: 0
    }
  );
  return u.current && v.observe(u.current), () => {
    u.current && v.unobserve(u.current);
  };
}, [f]), /* @__PURE__ */ _.jsx(
  "div",
  {
    ref: u,
    className: "infinite-list-item infinite-pull-down-refesh",
    "data-id": h,
    children: g || /* @__PURE__ */ _.jsx(Qe, { hasMore: f })
  }
)), wr = ({
  topLoadMoreCallback: u,
  bottomLoadMoreCallback: h,
  listConDomRef: w,
  hasMoreTopData: f,
  hasMoreBottomData: g,
  isFirstRender: v,
  oldHeight: m,
  start: F,
  setStart: R,
  index: b,
  item: y,
  updatePositions: C,
  children: P,
  loaderAtTop: O,
  loaderAtBottom: k,
  len: q
}) => {
  const x = $(null), L = $(!1), p = $(null), j = b === 1 && v.current;
  Ze(() => {
    G();
  }, []), le(() => {
    C(), re();
  }, [y]);
  const re = () => {
    (!L || p.current !== y) && K(), p.current = y, L.current = !0;
  }, K = () => {
    if (!x.current || !w.current) return;
    const z = x.current.getBoundingClientRect().height - m;
    z && b <= F && (w.current.scrollTop = w.current.scrollTop + z);
  }, G = () => {
    var I;
    j && (v.current = !1, R(1), (I = x.current) == null || I.scrollIntoView());
  };
  return b === 0 ? /* @__PURE__ */ _.jsx(
    Sr,
    {
      refs: x,
      dataId: b,
      topLoadMoreCallback: u,
      hasMoreTopData: f,
      loaderAtTop: O
    },
    b
  ) : b === q - 1 ? /* @__PURE__ */ _.jsx(
    Tr,
    {
      refs: x,
      dataId: b,
      bottomLoadMoreCallback: h,
      hasMoreBottomData: g,
      loaderAtBottom: k
    },
    b
  ) : /* @__PURE__ */ _.jsx("div", { ref: x, className: "infinite-list-item", "data-id": b, children: typeof P == "function" ? P({ item: y, index: b - 1 }) : P }, b);
}, xr = (u, h) => {
  let w = 0, f = u.length - 1, g = null;
  for (; w <= f; ) {
    let v = parseInt(((w + f) / 2).toString()), m = u[v].bottom;
    if (m === h)
      return v + 1;
    m < h ? w = v + 1 : m > h && ((g === null || g > v) && (g = v), f = f - 1);
  }
  return g;
};
function jr(u) {
  return Le.forwardRef(u);
}
let Xe = 0, Fe = 0;
const Pr = jr((u, h) => {
  const {
    // 列表数据
    listData: w = [],
    // 预估的列表项高度
    estimatedItemSize: f = 40,
    // 缓冲区的比例
    bufferScale: g = 1,
    children: v,
    pullDownCallback: m,
    pullUpCallback: F,
    hasMoreTopData: R = !0,
    hasMoreBottomData: b = !0,
    loaderAtTop: y,
    loaderAtBottom: C
  } = u, [P, O] = ee(0), [k, q] = ee(0), [x, L] = ee([{ id: "267e90" }, ...w, { id: "23ff8" }]), p = (a) => new Array(a || x.length).fill(!0).map((S, s) => ({
    index: s,
    height: f,
    top: s * f,
    bottom: (s + 1) * f
  })), [j, re] = ee(p), K = $(!1), G = $(!1), I = $(null), z = $(null), be = $(!0), B = $(""), ye = $(0), te = $(j), M = j[j.length - 1].bottom, W = Math.ceil(P / f), J = k + W, ce = Math.floor(g * W), ne = Math.min(k, ce), Re = Math.min(x.length - J, ce), Ee = j[k - ne].top, _e = x.slice(k - ne, J + Re);
  Rr(
    h,
    () => ({
      // 首尾是加载更多，中间才是真实的listData
      getListData() {
        return [...x.slice(1, x.length - 1)];
      },
      // 单个删除
      delItem(a) {
        if (a < 0 || a >= x.length - 2) return;
        const o = [...x], S = a + 1;
        o.splice(S, 1);
        const s = [...j], V = j[S].height;
        s.splice(S, 1);
        for (let Y = S; Y < s.length; Y++)
          s[Y].top = s[Y - 1].bottom, s[Y].bottom = s[Y].bottom - V, s[Y].index = Y;
        re(s), te.current = s, L(o);
      }
    }),
    [x]
  );
  const X = (a) => {
    if (!z.current) return;
    const o = a || j;
    let S = z.current.children;
    if (S && S.length > 0) {
      const s = [...o];
      Array.from(S).forEach((V) => {
        let ve = V.getBoundingClientRect().height, H = +V.dataset.id, pe = o[H].height - ve;
        if (pe) {
          s[H].bottom = s[H].bottom - pe, s[H].height = ve;
          for (let U = H + 1; U < s.length; U++)
            s[U].top = s[U - 1].bottom, s[U].bottom = s[U].bottom - pe;
        }
      }), re(s), te.current = s;
    }
  }, Se = (a = 0) => xr(j, a);
  le(() => {
    const a = I.current, o = je();
    return o == null || o.observe(a), () => {
      o == null || o.unobserve(a);
    };
  }, []), le(() => {
    X();
  }, [k]);
  const Te = async () => {
    if (G.current || !F) return;
    G.current = !0;
    const a = await F();
    G.current = !1;
    const o = ye.current;
    a.length !== 0 && (Fe = te.current[o].bottom - I.current.scrollTop, B.current = "recoverBottomScrollY", we(a));
  }, we = (a) => {
    L((o) => (X(p(o.length + a.length)), [...o.slice(0, o.length - 1), ...a, o[o.length - 1]]));
  };
  le(() => {
    xe();
  }, [x]);
  const xe = () => {
    B.current && (B.current === "recoverTopScrollY" && ue(), B.current === "recoverBottomScrollY" && fe(), B.current = "");
  }, ue = () => {
    const a = k + Xe;
    q(a);
    const o = j[k].bottom - I.current.scrollTop;
    I.current.scrollTop = j[a].bottom - o;
  }, fe = () => {
    I.current.scrollTop = j[k].bottom - Fe, Fe = 0;
  }, oe = async () => {
    if (K.current || !m) return;
    K.current = !0;
    const a = await m();
    K.current = !1, Xe = a.length, B.current = "recoverTopScrollY", de(a);
  }, de = (a) => {
    L((o) => (X(p(o.length + a.length)), [o[0], ...a, ...o.slice(1)]));
  }, Z = () => {
    let a = I.current.scrollTop;
    const o = Se(a);
    o !== null && (q(o), ye.current = o);
  }, je = () => {
    if (!("ResizeObserver" in window)) {
      O(I.current.clientHeight), console.error("浏览器不支持ResizeObserver，请pollyfill！");
      return;
    }
    return new ResizeObserver((o) => {
      const S = o[0], s = S.contentBoxSize ? S.contentBoxSize[0].blockSize : S.contentRect.height;
      O(s), Z();
    });
  };
  return /* @__PURE__ */ _.jsxs("div", { ref: I, className: "infinite-list-container", onScroll: Z, children: [
    /* @__PURE__ */ _.jsx("div", { className: "infinite-list-phantom", style: { height: M + "px" } }),
    /* @__PURE__ */ _.jsx("div", { ref: z, className: "infinite-list", style: { transform: `translate3d(0, ${Ee}px, 0)` }, children: _e.map((a, o) => {
      var s;
      const S = (s = j[o + k - ne]) == null ? void 0 : s.index;
      return S === void 0 ? null : /* @__PURE__ */ _.jsx(
        wr,
        {
          bottomLoadMoreCallback: Te,
          topLoadMoreCallback: oe,
          listConDomRef: I,
          hasMoreTopData: R,
          hasMoreBottomData: b,
          isFirstRender: be,
          oldHeight: j[S].height,
          start: k,
          index: S,
          item: a,
          updatePositions: X,
          children: v,
          loaderAtTop: y,
          loaderAtBottom: C,
          len: x.length,
          setStart: q
        },
        S
      );
    }) })
  ] });
});
export {
  Cr as FixedSizeList,
  Pr as VariableSizeList
};
